### PR TITLE
refactor(2918): subsquid for ProfileDetail

### LINKED ISSUE #2918: use 'subsquid' query in profile tabs
Currently we're using [magick](https://github.com/kodadot/magick) query in profile tabs, I would like them to be rewritten to use [subsquid](https://github.com/kodadot/rubick) for rmrk chain.

List of query:
- nftListByIssuer
- nftListCollected
- nftListSold
- firstNftByIssuer

### PR DESCRIPTION
**Pull request by KngZhi**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

### What's new?

- [x] PR closes #2918
- [ ] <brief_description_of_what_I've_added>

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

### Optional

- [ ] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [ ] I've tested PR on mobile and everything seems works
- [ ] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer?target=DaWoVNNhBTdqJ5iPN3UcaLeamkPpJeGWdrs8LhvX9TQbadc)

### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

### CODE REVIEW

**Comment by vikiival** [comment]
@KngZhi 
https://github.com/kodadot/nft-gallery/issues/2918#issuecomment-1120378565

**Comment by KngZhi** [comment]
> @KngZhi 
> 
> https://github.com/kodadot/nft-gallery/issues/2918#issuecomment-1120378565
> 
> 

okay, out of computer, will update soon.

File: components/rmrk/Gallery/PaginatedCardList.vue [hunk 1]
```diff
@@ -139,22 +156,22 @@ export default class PaginatedCardList extends mixins(
...
-      const { nodes, totalCount } = data.nFTEntities
-      const newNfts = nodes.map((e: any) => ({
-        ...e,
-        emoteCount: e?.emotes?.totalCount,
-      }))
-      await getCloudflareImageLinks(newNfts.map(mapOnlyMetadata)).catch(
+      const {
+        nftEntities,
+        nftEntitiesConnection: { totalCount },
+      } = data
```
Comment (roiLeo) [inline components/rmrk/Gallery/PaginatedCardList.vue:162]: where emotes?

File: components/rmrk/Gallery/PaginatedCardList.vue [hunk 2]
```diff
@@ -123,13 +136,17 @@ export default class PaginatedCardList extends mixins(
...
+      loadingKey: 'isLoading',
+      result: this.handleResult,
       variables: {
         account: this.account,
-        orderBy: this.searchQuery.sortBy,
-        search: this.buildSearchParam(),
-        first: this.first,
-        offset: (page - 1) * this.first,
+        orderBy: this.remapSortBy,
+        AND: this.buildSearchParam(),
```
Comment (roiLeo) [inline components/rmrk/Gallery/PaginatedCardList.vue:147]: I don't think `AND` will work

**Review by roiLeo — CHANGES_REQUESTED** [review]
- searchbar in profile don't seem to work
- your changes in `PaginatedCardList` could break other chain (statemin, westmint)
- could you add fragments in your queries? it will make it easier to read
- where emotes? added in #2936

**Comment by vikiival** [comment]
TBH I would just close both :)

**Comment by KngZhi** [comment]
@roiLeo 
> 1. searchbar in profile don't seem to work

Fixed with that. But another issue come up, *subsquid* seems does not to support **case-insensitive** name query.

*Rubick* totalCount 8, you will get another 4 counts for `L` name query
image

*Magic* totalCount 12
image

**Comment by roiLeo** [comment]
> @roiLeo
> 
> > 1. searchbar in profile don't seem to work
> 
> Fixed with that. But another issue come up, _subsquid_ seems does not to support **case-insensitive** name query.

#2951 might help you

**Comment by KngZhi** [comment]
> > @roiLeo
> 
> > 
> 
> > > 1. searchbar in profile don't seem to work
> 
> > 
> 
> > Fixed with that. But another issue come up, _subsquid_ seems does not to support **case-insensitive** name query.
> 
> 
> 
> #2951 might help you
> 
> 

Beautiful, many thanks!!

File: components/rmrk/Gallery/PaginatedCardList.vue [hunk 3]
```diff
@@ -138,7 +136,7 @@ export default class PaginatedCardList extends mixins(
       query: this.query,
       client: 'subsquid',
       variables: {
-        account: this.account,
+        account: 'Fksmad33PFxhrQXNYPPJozgWrv82zuFLvXK7Rh8m1xQhe98',
```
Comment (vikiival) [inline components/rmrk/Gallery/PaginatedCardList.vue:139]: 👀

**Review by vikiival — CHANGES_REQUESTED** [review]

File: components/rmrk/Gallery/PaginatedCardList.vue (hunk 3 above)
Comment (KngZhi) [reply components/rmrk/Gallery/PaginatedCardList.vue]: hahahahahahaha 😄,
I really dont have any nft yet, sorry, dear graphQL Mentor~

**Review by roiLeo — CHANGES_REQUESTED** [review]
Something went wrong when I test on my profil
Capture d’écran 2022-05-13 à 11 08 47 AM

**Comment by KngZhi** [comment]
local dev works fine, really confused 
@vikiival do you have any clue about this?
Screen Shot 2022-05-13 at 10 32 59 PM

**Comment by vikiival** [comment]
> local dev works fine, really confused @vikiival do you have any clue about this?

Maybe pass the query? :)

**Comment by KngZhi** [comment]
> > local dev works fine, really confused @vikiival do you have any clue about this?
> 
> Maybe pass the query? :)

sorry, don't understand, do you mind expanding a little more.

**Comment by vikiival** [comment]
This is the diff between 005 and 006
https://github.com/kodadot/rubick/compare/d14cd985822d4fc386c90228e71c8bd49865f61d..2820df7b1eadb3f11e138c0a931add95dbf6ba27

**Comment by vikiival** [comment]
I passed it to the subsquid team :)

**Comment by vikiival** [comment]
Updated 006 with new packages, should be workable in the couple of minutes

File: pages/westmint/u/_id.vue [hunk 4]
```diff
@@ -139,12 +139,13 @@ import isShareMode from '@/utils/isShareMode'
 import shouldUpdate from '@/utils/shouldUpdate'
 import shortAddress from '@/utils/shortAddress'
 import collectionList from '@/queries/unique/collectionListByAccount.graphql'
-import nftListByIssuer from '@/queries/unique/nftListByIssuer.graphql'
-import nftListCollected from '@/queries/unique/nftListCollected.graphql'
-import nftListSold from '@/queries/unique/nftListSold.graphql'
-import firstNftByIssuer from '@/queries/unique/firstNftByIssuer.graphql'
 import PrefixMixin from '@/utils/mixins/prefixMixin'
```
Comment (vikiival) [inline pages/westmint/u/_id.vue:143]: [:)](https://github.com/kodadot/nft-gallery/issues/2918#issuecomment-1120378565)

File: queries/fragments/rubickNft.graphql [hunk 5]
```diff
@@ -0,0 +1,11 @@
+fragment rubickNft on NFTEntity {
```
Comment (vikiival) [inline queries/fragments/rubickNft.graphql:1]: nah this deserves better name. (but it's not the focus of this PR)

**Review by vikiival — CHANGES_REQUESTED** [review]
Please make a proper path to the queries and we are good to go https://github.com/kodadot/nft-gallery/issues/2918#issuecomment-1120378565

File: pages/westmint/u/_id.vue (hunk 4 above)
Comment (KngZhi) [reply pages/westmint/u/_id.vue:143]: Oh, I see, you are saying be exactly like `'@/queries/subsquid/general/firstNftByIssuer.graphql'`

File: queries/fragments/rubickNft.graphql (hunk 5 above)
Comment (KngZhi) [reply queries/fragments/rubickNft.graphql:1]: subsquidNft? ...

Comment (vikiival) [reply queries/fragments/rubickNft.graphql:1]: Let's go with `subsquidNft` for now

**Review by roiLeo — COMMENTED** [review]
I can now see my nfts!

**Comment by KngZhi** [comment]
Out of computer, I will check it later.

By the way, thanks for testing!

**Comment by KngZhi** [comment]
> Don't know if this is related with this PR but History tab in user Profil don't show any amount

the beta website had the same issue.

image

**Review by roiLeo — APPROVED** [review]
✅ tested on preview, code lgtm

**Comment by KngZhi** [comment]
Wait a moment, I'm trying to fix with that.

**Comment by KngZhi** [comment]
😂

**Review by roiLeo — APPROVED** [review]
PR closes #3027

File: components/rmrk/Gallery/PaginatedCardList.vue [hunk 6]
```diff
@@ -123,14 +134,15 @@ export default class PaginatedCardList extends mixins(
     this.isFetchingData = true
     const result = await this.$apollo.query({
       query: this.query,
-      client: this.urlPrefix,
+      client: 'subsquid',
```
Comment (vikiival) [inline components/rmrk/Gallery/PaginatedCardList.vue:137]: @roiLeo would recommend `this.client`

File: components/rmrk/Gallery/PaginatedCardList.vue [hunk 7]
```diff
@@ -123,14 +134,15 @@ export default class PaginatedCardList extends mixins(
...
         account: this.account,
-        orderBy: this.searchQuery.sortBy,
-        search: this.buildSearchParam(),
-        first: this.first,
+        orderBy: this.remapSortBy,
+        and: this.buildSearchParam,
+        limit: this.first,
         offset: (page - 1) * this.first,
       },
+      fetchPolicy: 'no-cache',
```
Comment (vikiival) [inline components/rmrk/Gallery/PaginatedCardList.vue:145]: ?

Comment (KngZhi) [reply components/rmrk/Gallery/PaginatedCardList.vue:145]: Apollo diff bug.
if not using 'no-cache', it would return the same data.

**Comment by vikiival** [comment]
If Leo said LGTM we are good to go then. 

I need this page for Basilisk anyway :)

**Review by roiLeo — COMMENTED** [review]
@KngZhi 
I did a final test before merging and found that it was not possible to search in user "Sold" tab, same with "BUY NOW" filter. Sorting is working.
Capture d’écran 2022-05-17 à 2 03 08 PM

could you take a look?

**Comment by KngZhi** [comment]
> @KngZhi 
> 
> I did a final test before merging and found that it was not possible to search in user "Sold" tab, same with "BUY NOW" filter. Sorting is working.
> 
> Capture d’écran 2022-05-17 à 2 03 08 PM
> 
> 
> 
> could you take a look?

hmmm, okay! I will check that.

**Comment by KngZhi** [comment]
deployed version

https://user-images.githubusercontent.com/11361147/168840338-253b00a3-7311-4ea9-8c54-60047a98883c.mov

**Review by roiLeo — COMMENTED** [review]
Missing emotes added in #2936 & sort by emotes count

#### Beta
Capture d’écran 2022-05-18 à 10 54 41 AM

#### Preview
Capture d’écran 2022-05-18 à 10 55 12 AM

**Comment by KngZhi** [comment]
🫡 I'll check every corner, sir.

**Comment by KngZhi** [comment]
for the current subsquid, it does not have the field `emotes: { totalCount }` therefore it was unable to sortBy with that, we have to extend or found a workaround.

@vikiival any idea for how to do that? I check with subsquid official doc and magic repo, haven't found any clue.

**Comment by vikiival** [comment]
I would skip that for now :)

File: components/rmrk/Gallery/PaginatedCardList.vue [hunk 8]
```diff
@@ -74,19 +74,17 @@ export default class PaginatedCardList extends mixins(
   protected items: NFTWithMeta[] = []
   private isLoading = true
 
-  private buildSearchParam(): Record<string, unknown>[] {
+  private get buildSearchParam(): Record<string, unknown>[] {
```
Comment (vikiival) [inline components/rmrk/Gallery/PaginatedCardList.vue:77]: private getter 👀

File: components/rmrk/Gallery/History.vue [hunk 9]
```diff
@@ -275,6 +275,7 @@ export default class History extends mixins(ChainMixin, KeyboardEventsMixin) {
         priceCollectorMap[nftId] = '0'
       } else if (newEvent['interaction'] === 'BUY') {
         event['Type'] = this.$t('nft.event.BUY')
+        priceCollectorMap[nftId] = newEvent['meta']
```
Comment (vikiival) [inline components/rmrk/Gallery/History.vue:278]: Why we need this?

File: components/rmrk/Profile/ProfileDetail.vue [hunk 10]
```diff
@@ -426,7 +427,7 @@ export default class Profile extends mixins(
 
       this.$apollo.addSmartQuery('firstNft', {
         query: firstNftByIssuer,
-        client: this.urlPrefix,
+        client: 'subsquid',
```
Comment (vikiival) [inline components/rmrk/Profile/ProfileDetail.vue:430]:
```suggestion
        client: this.client,
```

**Review by vikiival — CHANGES_REQUESTED** [review]
smol stuff

File: components/rmrk/Gallery/History.vue (hunk 9 above)
Comment (KngZhi) [reply components/rmrk/Gallery/History.vue:278]: solve '-' issue. #3027

**Comment by KngZhi** [comment]
> I would skip that for now :)

if so, I would suggest we comment that option out for now.

File: pages/statemine/u/_id.vue [hunk 11]
```diff
@@ -259,7 +260,7 @@ export default class Profile extends mixins(PrefixMixin) {
       this.$apollo.addSmartQuery('collections', {
         query: collectionList,
         manual: true,
-        client: this.urlPrefix,
+        client: 'subsquid',
```
Comment (vikiival) [inline pages/statemine/u/_id.vue:263]:
```suggestion
        client: this.client,
```

File: pages/statemine/u/_id.vue [hunk 12]
```diff
@@ -276,7 +277,7 @@ export default class Profile extends mixins(PrefixMixin) {
       this.$apollo.addSmartQuery('firstNft', {
         query: firstNftByIssuer,
         manual: true,
-        client: this.urlPrefix,
+        client: 'subsquid',
```
Comment (vikiival) [inline pages/statemine/u/_id.vue:280]:
```suggestion
        client: this.client,
```

File: pages/westmint/u/_id.vue [hunk 13]
```diff
@@ -276,7 +277,7 @@ export default class Profile extends mixins(PrefixMixin) {
       this.$apollo.addSmartQuery('firstNft', {
         query: firstNftByIssuer,
         manual: true,
-        client: this.urlPrefix,
+        client: 'subsquid',
```
Comment (vikiival) [inline pages/westmint/u/_id.vue:280]:
```suggestion
        client: this.client,
```

File: components/rmrk/Gallery/History.vue (hunk 9 above)
Comment (vikiival) [reply components/rmrk/Gallery/History.vue:278]: I do not understand this. Maybe ELI5?

**Review by vikiival — CHANGES_REQUESTED** [review]
Hope to merge this today 🤞

File: components/rmrk/Gallery/History.vue (hunk 9 above)
Comment (KngZhi) [reply components/rmrk/Gallery/History.vue:278]: Hmm, sorry, could elaborate ELI5 , it means?

**Comment by KngZhi** [comment]
Out of computer, will be back soon.

File: components/rmrk/Gallery/History.vue (hunk 9 above)
Comment (vikiival) [reply components/rmrk/Gallery/History.vue:278]: Explain it like I am 5-year-old person

Comment (KngZhi) [reply components/rmrk/Gallery/History.vue:278]: It's irrelevant with refactoring, just free ride to solve a bug roiLee found.

👌 the issue I mentioned before, it does not show the money.

The reason is we did not add money data to priceCollectorMap when event is 'buy'.

Comment (vikiival) [reply components/rmrk/Gallery/History.vue:278]: OK :D but still confused why we do not use meta from the buy event ?

Comment (KngZhi) [reply components/rmrk/Gallery/History.vue:278]: image

here is the original logic.

Comment (KngZhi) [reply components/rmrk/Gallery/History.vue:278]: if for `Buy` event we do not set `priceCollectorMap` for nft, it would return `undefined`, then the ternary would go to `'-'`

**Comment by KngZhi** [comment]
`this.client` does not support on test net, it would goes to `https://api.subquery.network/sq/kodadot/unique`.

**Comment by vikiival** [comment]
>  test net

:D Unique is implementation of Statemine, it's not a testnet
But do not worry about that :)

**Review by vikiival — APPROVED** [review]

**Comment by vikiival** [comment]
pay 250 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $250 @ 81.06 USD/KSM ~ 3.084 $KSM
🧗 [DaWoVNNhBTdqJ5iPN3UcaLeamkPpJeGWdrs8LhvX9TQbadc](https://kusama.subscan.io/account/DaWoVNNhBTdqJ5iPN3UcaLeamkPpJeGWdrs8LhvX9TQbadc)
🔗 [0x6e7ef84695d2bb4db3ead4e9af3c9bfe2de8069e76c500cfb4275da6cc3d8157](https://kusama.subscan.io/extrinsic/0x6e7ef84695d2bb4db3ead4e9af3c9bfe2de8069e76c500cfb4275da6cc3d8157)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)
